hls-test-utils/src/Test/Hls.hs

```diff
@@ -308,6 +391,90 @@ runSessionWithServerFormatter plugin formatter conf fp act = do
     fp
     act
 
+runSessionWithServerInTmpDir :: Pretty b => PluginTestDescriptor b -> VirtualFileTree -> Session a -> IO a
+runSessionWithServerInTmpDir plugin tree act = do
+  recorder <- pluginTestRecorder
+  runSessionWithServerInTmpDir' (plugin recorder) def def fullCaps tree act
+
+runSessionWithServerAndCapsInTmpDir :: Pretty b => PluginTestDescriptor b -> ClientCapabilities -> VirtualFileTree -> Session a -> IO a
+runSessionWithServerAndCapsInTmpDir plugin caps tree act = do
+  recorder <- pluginTestRecorder
+  runSessionWithServerInTmpDir' (plugin recorder) def def caps tree act
+
+runSessionWithServerFormatterInTmpDir :: Pretty b => PluginTestDescriptor b -> String -> PluginConfig ->  VirtualFileTree -> Session a -> IO a
+runSessionWithServerFormatterInTmpDir plugin formatter conf tree act = do
+  recorder <- pluginTestRecorder
+  runSessionWithServerInTmpDir'
+    (plugin recorder)
+    def
+      { formattingProvider = T.pack formatter
+      , plugins = M.singleton (PluginId $ T.pack formatter) conf
+      }
+    def
+    fullCaps
+    tree
+    act
+
+-- | Host a server, and run a test session on it.
+--
+-- Creates a temporary directory, and materializes the VirtualFileTree
+-- in the temporary directory.
+--
+-- To debug test cases and verify the file system is correctly set up,
+-- you should set the environment variable 'HLS_TEST_HARNESS_NO_TESTDIR_CLEANUP=1'.
+-- Further, we log the temporary directory location on startup. To view
+-- the logs, set the environment variable 'HLS_TEST_HARNESS_STDERR=1'.
```

Comment:
The environment variables are nice because they require no recompilation to take effect.  When you run `cabal test --test-options`, then any log level change causes recompilation. `cabal run <plugin-name>:tests -- <opts>` doesn't work from the HLS project root, you have to cd to the plugin directory. `cabal run` and `cabal test` set the working directory differently.